column_index.go

```diff
@@ -676,19 +676,29 @@ func truncateLargeMinByteArrayValue(value []byte, sizeLimit int) []byte {
 }
 
 func truncateLargeMaxByteArrayValue(value []byte, sizeLimit int) []byte {
-	if len(value) > sizeLimit && !isMaxByteArrayValue(value) {
-		value = value[:sizeLimit]
+	if len(value) > sizeLimit {
+		if newValue := incrementByteArray(value[:sizeLimit]); newValue != nil {
+			value = newValue
+		}
 	}
 	return value
 }
 
-func isMaxByteArrayValue(value []byte) bool {
-	for i := range value {
-		if value[i] != 0xFF {
-			return false
+// incrementByteArray increments the given byte array by 1.
+// Returns nil if the given byte array is all 0xFF.
+// Reference: https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/internal/column/columnindex/BinaryTruncator.java#L124
+func incrementByteArray(value []byte) []byte {
+	newValue := make([]byte, len(value))
+	copy(newValue, value)
+	for i := len(value) - 1; i >= 0; i-- {
+		elem := newValue[i]
+		elem++
+		newValue[i] = elem
+		if elem != 0 { // Did not overflow: 0xFF -> 0x00
+			return newValue
 		}
 	}
-	return true
+	return nil
 }
```

Comment:
Looking this over again, we should avoid the unnecessary allocations. And in the rare case that it fully overflowed we just correct it back.:
```
// incrementByteArray increments the given byte array by 1.
// Reference: https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/internal/column/columnindex/BinaryTruncator.java#L124
func incrementByteArray(value []byte) {
	for i := len(value) - 1; i >= 0; i-- {
		value[i]++
		if value[i] != 0 { // Did not overflow: 0xFF -> 0x00
			return
		}
	}
	// Fully overflowed, so restore all to 0xFF
	for i := range value {
		value[i] = 0xFF
	}
}
```

Reply:
Hi, thanks for commenting.

I looked into a few places, it seems the call stack would be like this, for ByteArrayColumn:
- [byteArrayColumnIndexer.IndexPage](https://github.com/segmentio/parquet-go/blob/main/column_index.go#L508): did not copy min, `min.byteArray()` is a cast
- [writerColumn.recordPageStats](https://github.com/segmentio/parquet-go/blob/main/writer.go#L1313): did not copy returned values from `page.Bounds()`
- [byteArrayPage.Bounds](https://github.com/segmentio/parquet-go/blob/main/page.go#L1037): using returned values from `page.bounds()` without copying, where returned values of `page.bounds()` are taken from `page.values` by reslicing, so no copy as well.

If I am not mistaken, for byteArrayColumnIndex, modifying the input slice in place would change the underlying buffer of `page.values` as well.  

If we want to eliminate memory alloc, I'd propose we do it like the way we deal with FixedLenByteArrayColumn, that is appending to `byteArrayColumnIndexer.minValues` and `byteArrayColumnIndexer.maxValues` first (which is effectively a copy), then modify that buffer in place.